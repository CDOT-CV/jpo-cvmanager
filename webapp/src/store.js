import { configureStore } from "@reduxjs/toolkit";
import rsuReducer from "./slices/rsuSlice";
import userReducer from "./slices/userSlice";
import wzdxReducer from "./slices/wzdxSlice";
import configReducer from "./slices/configSlice";

const store = configureStore({
  reducer: {
    rsu: rsuReducer,
    user: userReducer,
    wzdx: wzdxReducer,
    config: configReducer,
  },
});

export default store;
