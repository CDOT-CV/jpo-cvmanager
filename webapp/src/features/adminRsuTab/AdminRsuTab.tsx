import React, { useState, useEffect } from 'react'
import AdminAddRsu from '../adminAddRsu/AdminAddRsu'
import AdminEditRsu, { AdminEditRsuFormType } from '../adminEditRsu/AdminEditRsu'
import AdminTable from '../../components/AdminTable'
import { IoChevronBackCircleOutline, IoRefresh } from 'react-icons/io5'
import { AiOutlinePlusCircle } from 'react-icons/ai'
import { confirmAlert } from 'react-confirm-alert'
import { Options } from '../../components/AdminDeletionOptions'
import {
  selectLoading,
  selectActiveDiv,
  selectTableData,
  selectTitle,
  selectEditRsuRowData,

  // actions
  updateTableData,
  setTitle,
  setActiveDiv,
  deleteMultipleRsus,
  deleteRsu,
  setEditRsuRowData,
} from './adminRsuTabSlice'
import { useSelector, useDispatch } from 'react-redux'

import './Admin.css'
import { AnyAction, ThunkDispatch } from '@reduxjs/toolkit'
import { RootState } from '../../store'
import { Action } from '@material-table/core'
import { AdminRsu } from '../../types/Rsu'

const AdminRsuTab = () => {
  const dispatch: ThunkDispatch<RootState, void, AnyAction> = useDispatch()

  const activeDiv = useSelector(selectActiveDiv)
  const tableData = useSelector(selectTableData)
  const title = useSelector(selectTitle)
  const [columns] = useState([
    { title: 'Milepost', field: 'milepost', id: 0 },
    { title: 'IP Address', field: 'ip', id: 1 },
    { title: 'Primary Route', field: 'primary_route', id: 2 },
    { title: 'RSU Model', field: 'model', id: 3 },
    { title: 'Serial Number', field: 'serial_number', id: 4 },
  ])
  const editRsuRowData = useSelector(selectEditRsuRowData)

  const loading = useSelector(selectLoading)

  const tableActions: Action<AdminEditRsuFormType>[] = [
    {
      icon: 'delete',
      tooltip: 'Delete RSU',
      position: 'row',
      onClick: (event, rowData: AdminEditRsuFormType) => {
        const buttons = [
          { label: 'Yes', onClick: () => onDelete(rowData) },
          { label: 'No', onClick: () => {} },
        ]
        const alertOptions = Options('Delete RSU', 'Are you sure you want to delete "' + rowData.ip + '"?', buttons)
        confirmAlert(alertOptions)
      },
    },
    {
      icon: 'edit',
      tooltip: 'Edit RSU',
      position: 'row',
      onClick: (event, rowData: AdminEditRsuFormType) => onEdit(rowData),
    },
    {
      tooltip: 'Remove All Selected From Organization',
      icon: 'delete',
      onClick: (event, rowData: AdminEditRsuFormType[]) => {
        const buttons = [
          { label: 'Yes', onClick: () => multiDelete(rowData) },
          { label: 'No', onClick: () => {} },
        ]
        const alertOptions = Options(
          'Delete Selected RSUs',
          'Are you sure you want to delete ' + rowData.length + ' RSUs?',
          buttons
        )
        confirmAlert(alertOptions)
      },
    },
  ]
  useEffect(() => {
    dispatch(setActiveDiv('rsu_table'))
  }, [dispatch])

  useEffect(() => {
    dispatch(setTitle())
  }, [activeDiv, dispatch])

  const onEdit = (row: AdminEditRsuFormType) => {
    dispatch(setEditRsuRowData(row))
    dispatch(setActiveDiv('edit_rsu'))
  }

  const onDelete = (row: AdminEditRsuFormType) => {
    dispatch(deleteRsu({ rsu_ip: row.ip, shouldUpdateTableData: true }))
  }

  const multiDelete = (rows: AdminEditRsuFormType[]) => {
    dispatch(deleteMultipleRsus(rows))
  }

  return (
    <div>
      <div>
        <h3 className="panel-header">
          {activeDiv !== 'rsu_table' && (
            <button
              key="rsu_table"
              className="admin_table_button"
              onClick={(value) => {
                dispatch(setActiveDiv('rsu_table'))
              }}
            >
              <IoChevronBackCircleOutline size={20} />
            </button>
          )}
          {title}
          {activeDiv === 'rsu_table' && [
            <button
              key="plus_button"
              className="plus_button"
              onClick={(value) => {
                dispatch(setActiveDiv('add_rsu'))
              }}
              title="Add RSU"
            >
              <AiOutlinePlusCircle size={20} />
            </button>,
            <button
              key="refresh_button"
              className="plus_button"
              onClick={(value) => {
                dispatch(updateTableData())
              }}
              title="Refresh RSUs"
            >
              <IoRefresh size={20} />
            </button>,
          ]}
        </h3>
      </div>
      {activeDiv === 'rsu_table' && loading === false && (
        <div className="scroll-div-tab">
          <AdminTable title={''} data={tableData} columns={columns} actions={tableActions} />
        </div>
      )}

      {activeDiv === 'add_rsu' && (
        <div className="scroll-div-tab">
          <AdminAddRsu />
        </div>
      )}

      {activeDiv === 'edit_rsu' && (
        <div className="scroll-div-tab">
          <AdminEditRsu rsuData={editRsuRowData} />
        </div>
      )}
    </div>
  )
}

export default AdminRsuTab
