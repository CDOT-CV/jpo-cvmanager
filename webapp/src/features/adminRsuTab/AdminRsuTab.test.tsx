import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import AdminRsuTab from './AdminRsuTab'
import { Provider } from 'react-redux'
import { setupStore } from '../../store'
import { replaceChaoticIds } from '../../utils/test-utils'
import { BrowserRouter } from 'react-router-dom'

it('snapshot add', () => {
  const { container } = render(
    <Provider store={setupStore({ adminRsuTab: { loading: false, value: { activeDiv: 'rsu_table' } } })}>
      <BrowserRouter>
        <AdminRsuTab />
      </BrowserRouter>
    </Provider>
  )

  fireEvent.click(screen.queryByTitle('Add RSU'))

  expect(replaceChaoticIds(container)).toMatchSnapshot()
})

it('snapshot refresh', () => {
  const { container } = render(
    <Provider store={setupStore({ adminRsuTab: { loading: false, value: { activeDiv: 'rsu_table' } } })}>
      <BrowserRouter>
        <AdminRsuTab />
      </BrowserRouter>
    </Provider>
  )

  fireEvent.click(screen.queryByTitle('Refresh RSUs'))

  expect(replaceChaoticIds(container)).toMatchSnapshot()
})
