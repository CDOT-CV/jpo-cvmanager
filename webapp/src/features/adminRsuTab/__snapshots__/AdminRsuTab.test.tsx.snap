// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot add 1`] = `
<div>
  <div>
    <div>
      <h3
        class="panel-header"
      >
        CV Manager RSUs
        <button
          class="plus_button"
          title="Add RSU"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
            />
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
            />
          </svg>
        </button>
        <button
          class="plus_button"
          title="Refresh RSUs"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320 146s24.36-12-64-12a160 160 0 10160 160"
              fill="none"
              stroke-linecap="round"
              stroke-miterlimit="10"
              stroke-width="32"
            />
            <path
              d="M256 58l80 80-80 80"
              fill="none"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="32"
            />
          </svg>
        </button>
      </h3>
    </div>
    <div
      class="scroll-div-tab"
    >
      <div>
        <form
          class=""
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="ip"
            >
              RSU IP
            </label>
            <input
              class="form-control"
              id="ip"
              name="ip"
              placeholder="Enter RSU IP (Required)"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="latitude"
            >
              Latitude
            </label>
            <input
              class="form-control"
              id="latitude"
              name="latitude"
              placeholder="Enter RSU Latitude (Required)"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="longitude"
            >
              Longitude
            </label>
            <input
              class="form-control"
              id="longitude"
              name="longitude"
              placeholder="Enter RSU Longitude (Required)"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="milepost"
            >
              Milepost
            </label>
            <input
              class="form-control"
              id="milepost"
              name="milepost"
              placeholder="Enter RSU Milepost (Required)"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="primary_route"
            >
              Primary Route
            </label>
            <div
              aria-autocomplete="list"
              aria-busy="false"
              aria-controls="rw_2_listbox"
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="rw_2_listbox"
              aria-readonly="false"
              class="form-dropdown rw-dropdown-list rw-widget"
              data-intent="mouse"
              id="rw_1_input"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="rw-widget-input rw-widget-picker rw-widget-container"
                tabindex="-1"
              >
                <div
                  class="rw-dropdown-list-input"
                >
                  <input
                    aria-hidden="true"
                    class="rw-detect-autofill rw-sr"
                    tabindex="-1"
                    value="Select Route (Required)"
                  />
                  <input
                    autocomplete="off"
                    class="rw-dropdownlist-search"
                    size="2"
                    value=""
                  />
                  <span
                    class="rw-dropdown-list-value"
                  >
                    Select Route (Required)
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="rw-btn rw-picker-caret"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentcolor"
                    height="1em"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="serial_number"
            >
              Serial Number
            </label>
            <input
              class="form-control"
              id="serial_number"
              name="serial_number"
              placeholder="Enter RSU Serial Number (Required)"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="model"
            >
              RSU Model
            </label>
            <div
              aria-autocomplete="list"
              aria-busy="false"
              aria-controls="rw_5_listbox"
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="rw_5_listbox"
              aria-readonly="false"
              class="form-dropdown rw-dropdown-list rw-widget"
              data-intent="mouse"
              id="rw_4_input"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="rw-widget-input rw-widget-picker rw-widget-container"
                tabindex="-1"
              >
                <div
                  class="rw-dropdown-list-input"
                >
                  <input
                    aria-hidden="true"
                    class="rw-detect-autofill rw-sr"
                    tabindex="-1"
                    value="Select RSU Model (Required)"
                  />
                  <input
                    autocomplete="off"
                    class="rw-dropdownlist-search"
                    size="2"
                    value=""
                  />
                  <span
                    class="rw-dropdown-list-value"
                  >
                    Select RSU Model (Required)
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="rw-btn rw-picker-caret"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentcolor"
                    height="1em"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="scms_id"
            >
              SCMS ID
            </label>
            <input
              class="form-control"
              id="scms_id"
              name="scms_id"
              placeholder="Enter SCMS ID (Required)"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="ssh_credential_group"
            >
              SSH Credential Group
            </label>
            <div
              aria-autocomplete="list"
              aria-busy="false"
              aria-controls="rw_8_listbox"
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="rw_8_listbox"
              aria-readonly="false"
              class="form-dropdown rw-dropdown-list rw-widget"
              data-intent="mouse"
              id="rw_7_input"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="rw-widget-input rw-widget-picker rw-widget-container"
                tabindex="-1"
              >
                <div
                  class="rw-dropdown-list-input"
                >
                  <input
                    aria-hidden="true"
                    class="rw-detect-autofill rw-sr"
                    tabindex="-1"
                    value="Select SSH Group (Required)"
                  />
                  <input
                    autocomplete="off"
                    class="rw-dropdownlist-search"
                    size="2"
                    value=""
                  />
                  <span
                    class="rw-dropdown-list-value"
                  >
                    Select SSH Group (Required)
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="rw-btn rw-picker-caret"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentcolor"
                    height="1em"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="snmp_credential_group"
            >
              SNMP Credential Group
            </label>
            <div
              aria-autocomplete="list"
              aria-busy="false"
              aria-controls="rw_11_listbox"
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="rw_11_listbox"
              aria-readonly="false"
              class="form-dropdown rw-dropdown-list rw-widget"
              data-intent="mouse"
              id="rw_10_input"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="rw-widget-input rw-widget-picker rw-widget-container"
                tabindex="-1"
              >
                <div
                  class="rw-dropdown-list-input"
                >
                  <input
                    aria-hidden="true"
                    class="rw-detect-autofill rw-sr"
                    tabindex="-1"
                    value="Select SNMP Group (Required)"
                  />
                  <input
                    autocomplete="off"
                    class="rw-dropdownlist-search"
                    size="2"
                    value=""
                  />
                  <span
                    class="rw-dropdown-list-value"
                  >
                    Select SNMP Group (Required)
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="rw-btn rw-picker-caret"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentcolor"
                    height="1em"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="snmp_version_group"
            >
              SNMP Version
            </label>
            <div
              aria-autocomplete="list"
              aria-busy="false"
              aria-controls="rw_14_listbox"
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="rw_14_listbox"
              aria-readonly="false"
              class="form-dropdown rw-dropdown-list rw-widget"
              data-intent="mouse"
              id="rw_13_input"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="rw-widget-input rw-widget-picker rw-widget-container"
                tabindex="-1"
              >
                <div
                  class="rw-dropdown-list-input"
                >
                  <input
                    aria-hidden="true"
                    class="rw-detect-autofill rw-sr"
                    tabindex="-1"
                    value="Select SNMP Version (Required)"
                  />
                  <input
                    autocomplete="off"
                    class="rw-dropdownlist-search"
                    size="2"
                    value=""
                  />
                  <span
                    class="rw-dropdown-list-value"
                  >
                    Select SNMP Version (Required)
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="rw-btn rw-picker-caret"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentcolor"
                    height="1em"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="organizations"
            >
              Organization
            </label>
            <div
              class="form-dropdown rw-multiselect rw-widget"
              data-intent="mouse"
              tabindex="-1"
            >
              <div
                class="rw-widget-input rw-widget-picker rw-widget-container"
              >
                <div
                  aria-label="Selected items"
                  aria-multiselectable="true"
                  aria-orientation="horizontal"
                  class="rw-multiselect-taglist"
                  id="rw_17_taglist"
                  role="listbox"
                >
                  <input
                    aria-autocomplete="list"
                    aria-busy="false"
                    aria-controls="rw_18_listbox"
                    aria-disabled="false"
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-owns="rw_18_listbox "
                    aria-readonly="false"
                    autocapitalize="off"
                    autocomplete="off"
                    class="rw-multiselect-input"
                    placeholder="Select Organizations (Required)"
                    role="combobox"
                    size="32"
                    spellcheck="false"
                    tabindex="0"
                    value=""
                  />
                </div>
                <span
                  aria-hidden="true"
                  class="rw-btn rw-picker-caret"
                />
              </div>
            </div>
          </div>
          <p
            class="success-msg"
          />
          <p
            class="error-msg"
          >
            Failed to add rsu due to error: 
            
          </p>
          <div
            class="form-control"
          >
            <label />
            <button
              class="admin-button"
              type="submit"
            >
              Add RSU
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`snapshot refresh 1`] = `
<div>
  <div>
    <div>
      <h3
        class="panel-header"
      >
        CV Manager RSUs
        <button
          class="plus_button"
          title="Add RSU"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
            />
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
            />
          </svg>
        </button>
        <button
          class="plus_button"
          title="Refresh RSUs"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320 146s24.36-12-64-12a160 160 0 10160 160"
              fill="none"
              stroke-linecap="round"
              stroke-miterlimit="10"
              stroke-width="32"
            />
            <path
              d="M256 58l80 80-80 80"
              fill="none"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="32"
            />
          </svg>
        </button>
      </h3>
    </div>
    <div
      class="scroll-div-tab"
    >
      <div>
        <form
          class=""
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="ip"
            >
              RSU IP
            </label>
            <input
              class="form-control"
              id="ip"
              name="ip"
              placeholder="Enter RSU IP"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="latitude"
            >
              Latitude
            </label>
            <input
              class="form-control"
              id="latitude"
              name="latitude"
              placeholder="Enter RSU Latitude"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="longitude"
            >
              Longitude
            </label>
            <input
              class="form-control"
              id="longitude"
              name="longitude"
              placeholder="Enter RSU Longitude"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="milepost"
            >
              Milepost
            </label>
            <input
              class="form-control"
              id="milepost"
              name="milepost"
              placeholder="Enter RSU Milepost"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="primary_route"
            >
              Primary Route
            </label>
            <div
              aria-autocomplete="list"
              aria-busy="false"
              aria-controls="rw_23_listbox"
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="rw_23_listbox"
              aria-readonly="false"
              class="form-dropdown rw-dropdown-list rw-widget"
              data-intent="mouse"
              id="rw_22_input"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="rw-widget-input rw-widget-picker rw-widget-container"
                tabindex="-1"
              >
                <div
                  class="rw-dropdown-list-input"
                >
                  <input
                    aria-hidden="true"
                    class="rw-detect-autofill rw-sr"
                    tabindex="-1"
                    value="Select Route"
                  />
                  <input
                    autocomplete="off"
                    class="rw-dropdownlist-search"
                    size="2"
                    value=""
                  />
                  <span
                    class="rw-dropdown-list-value"
                  >
                    Select Route
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="rw-btn rw-picker-caret"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentcolor"
                    height="1em"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="serial_number"
            >
              Serial Number
            </label>
            <input
              class="form-control"
              id="serial_number"
              name="serial_number"
              placeholder="Enter RSU Serial Number"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="model"
            >
              RSU Model
            </label>
            <div
              aria-autocomplete="list"
              aria-busy="false"
              aria-controls="rw_26_listbox"
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="rw_26_listbox"
              aria-readonly="false"
              class="form-dropdown rw-dropdown-list rw-widget"
              data-intent="mouse"
              id="rw_25_input"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="rw-widget-input rw-widget-picker rw-widget-container"
                tabindex="-1"
              >
                <div
                  class="rw-dropdown-list-input"
                >
                  <input
                    aria-hidden="true"
                    class="rw-detect-autofill rw-sr"
                    tabindex="-1"
                    value="Select RSU Model"
                  />
                  <input
                    autocomplete="off"
                    class="rw-dropdownlist-search"
                    size="2"
                    value=""
                  />
                  <span
                    class="rw-dropdown-list-value"
                  >
                    Select RSU Model
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="rw-btn rw-picker-caret"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentcolor"
                    height="1em"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="scms_id"
            >
              SCMS ID
            </label>
            <input
              class="form-control"
              id="scms_id"
              name="scms_id"
              placeholder="Enter SCMS ID"
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="ssh_credential_group"
            >
              SSH Credential Group
            </label>
            <div
              aria-autocomplete="list"
              aria-busy="false"
              aria-controls="rw_29_listbox"
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="rw_29_listbox"
              aria-readonly="false"
              class="form-dropdown rw-dropdown-list rw-widget"
              data-intent="mouse"
              id="rw_28_input"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="rw-widget-input rw-widget-picker rw-widget-container"
                tabindex="-1"
              >
                <div
                  class="rw-dropdown-list-input"
                >
                  <input
                    aria-hidden="true"
                    class="rw-detect-autofill rw-sr"
                    tabindex="-1"
                    value="Select SSH Group"
                  />
                  <input
                    autocomplete="off"
                    class="rw-dropdownlist-search"
                    size="2"
                    value=""
                  />
                  <span
                    class="rw-dropdown-list-value"
                  >
                    Select SSH Group
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="rw-btn rw-picker-caret"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentcolor"
                    height="1em"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="snmp_credential_group"
            >
              SNMP Credential Group
            </label>
            <div
              aria-autocomplete="list"
              aria-busy="false"
              aria-controls="rw_32_listbox"
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="rw_32_listbox"
              aria-readonly="false"
              class="form-dropdown rw-dropdown-list rw-widget"
              data-intent="mouse"
              id="rw_31_input"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="rw-widget-input rw-widget-picker rw-widget-container"
                tabindex="-1"
              >
                <div
                  class="rw-dropdown-list-input"
                >
                  <input
                    aria-hidden="true"
                    class="rw-detect-autofill rw-sr"
                    tabindex="-1"
                    value="Select SNMP Group"
                  />
                  <input
                    autocomplete="off"
                    class="rw-dropdownlist-search"
                    size="2"
                    value=""
                  />
                  <span
                    class="rw-dropdown-list-value"
                  >
                    Select SNMP Group
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="rw-btn rw-picker-caret"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentcolor"
                    height="1em"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="snmp_version_group"
            >
              SNMP Version
            </label>
            <div
              aria-autocomplete="list"
              aria-busy="false"
              aria-controls="rw_35_listbox"
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="rw_35_listbox"
              aria-readonly="false"
              class="form-dropdown rw-dropdown-list rw-widget"
              data-intent="mouse"
              id="rw_34_input"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="rw-widget-input rw-widget-picker rw-widget-container"
                tabindex="-1"
              >
                <div
                  class="rw-dropdown-list-input"
                >
                  <input
                    aria-hidden="true"
                    class="rw-detect-autofill rw-sr"
                    tabindex="-1"
                    value="Select SNMP Version"
                  />
                  <input
                    autocomplete="off"
                    class="rw-dropdownlist-search"
                    size="2"
                    value=""
                  />
                  <span
                    class="rw-dropdown-list-value"
                  >
                    Select SNMP Version
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="rw-btn rw-picker-caret"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentcolor"
                    height="1em"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="organizations"
            >
              Organization
            </label>
            <div
              class="form-dropdown rw-multiselect rw-widget"
              data-intent="mouse"
              tabindex="-1"
            >
              <div
                class="rw-widget-input rw-widget-picker rw-widget-container"
              >
                <div
                  aria-label="Selected items"
                  aria-multiselectable="true"
                  aria-orientation="horizontal"
                  class="rw-multiselect-taglist"
                  id="rw_38_taglist"
                  role="listbox"
                >
                  <input
                    aria-autocomplete="list"
                    aria-busy="false"
                    aria-controls="rw_39_listbox"
                    aria-disabled="false"
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-owns="rw_39_listbox "
                    aria-readonly="false"
                    autocapitalize="off"
                    autocomplete="off"
                    class="rw-multiselect-input"
                    placeholder="Select organizations"
                    role="combobox"
                    size="21"
                    spellcheck="false"
                    tabindex="0"
                    value=""
                  />
                </div>
                <span
                  aria-hidden="true"
                  class="rw-btn rw-picker-caret"
                />
              </div>
            </div>
          </div>
          <p
            class="success-msg"
          />
          <p
            class="error-msg"
          >
            Failed to add rsu due to error: 
            
          </p>
          <div
            class="form-control"
          >
            <label />
            <button
              class="admin-button"
              type="submit"
            >
              Add RSU
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
